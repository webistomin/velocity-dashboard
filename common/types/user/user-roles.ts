export enum UserRoles {
  ADMIN = 'admin',
  OPERATOR = 'operator',
}
