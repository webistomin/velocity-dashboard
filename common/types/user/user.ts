import { IUserSchema } from './user-schema';

export interface IOwnUserProfileResponseBody {
  success: boolean;
  message?: string;
  user?: IUserSchema;
}
