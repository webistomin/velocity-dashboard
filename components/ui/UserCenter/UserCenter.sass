+b('user-center')
  +e('btn')
    width: rem(50px)
    height: rem(50px)
    max-width: rem(60px)
    display: flex
    align-items: center
    justify-content: center
