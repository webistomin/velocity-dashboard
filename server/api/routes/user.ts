import { Router, Request, Response } from 'express';
const route = Router();

export default (app: Router) => {
  app.use('/users', route);

  route.get('/me', (req: Request, res: Response) => {
    return res.json({ user: req }).status(200);
  });
};
