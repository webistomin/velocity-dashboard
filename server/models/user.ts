import mongoose, { Schema, HookNextFunction } from 'mongoose';
import bcrypt from 'bcrypt-nodejs';
import { IUserSchema } from 'common/types/user/user-schema';
import { SiteThemes } from 'common/types/theme/site-themes';

const UserSchema: Schema = new Schema({
  firstName: {
    type: String,
    required: true,
  },
  lastName: {
    type: String,
    required: true,
  },
  role: {
    type: String,
    required: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
  },
  password: {
    type: String,
    required: true,
  },
  theme: {
    type: String,
    default: SiteThemes.SHELOB,
  },
  notifications: {
    isEmailNotificationsEnabled: {
      type: Boolean,
      default: true,
    },
    isPushNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isMonthlyNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isQuarterNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
  },
  phone: {
    type: String,
    default: '',
  },
  socials: {
    twitter: {
      type: String,
      default: '',
    },
  },
  location: {
    type: String,
    default: '',
  },
  bio: {
    type: String,
    default: '',
  },
});

UserSchema.pre<IUserSchema>('save', function(next: HookNextFunction) {
  const user = this;
  if (this.isModified('password') || this.isNew) {
    bcrypt.genSalt(10, function(err, salt) {
      if (err) {
        return next(err);
      }

      bcrypt.hash(user.password, salt, null, function(err, hash) {
        if (err) {
          return next(err);
        }

        user.password = hash;
        next();
      });
    });
  } else {
    return next();
  }
});

UserSchema.methods.comparePassword = function(password: IUserSchema['password']): Boolean {
  const user = this;
  return bcrypt.compareSync(password, user.password);
};

export default mongoose.model<IUserSchema>('User', UserSchema);
