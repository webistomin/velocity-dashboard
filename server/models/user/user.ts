import { Schema, model, HookNextFunction } from 'mongoose';
import bcrypt from 'bcrypt';
import { nanoid } from 'nanoid';
import { SentMessageInfo } from 'nodemailer';
import consola from 'consola';

import { IUserInterface, IUserSchema } from 'common/types/user/user-schema';
import { SiteThemes } from 'common/types/theme/site-themes';
import PasswordReset from 'server/models/auth/password-reset';
import { sendResetMail } from 'server/services/mailer/reset-mail';

const UserSchema: Schema = new Schema<IUserInterface>({
  firstName: {
    type: String,
    required: true,
  },
  lastName: {
    type: String,
    required: true,
  },
  role: {
    type: String,
    required: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
  },
  password: {
    type: String,
    required: true,
  },
  theme: {
    type: String,
    default: SiteThemes.SHELOB,
  },
  notifications: {
    isEmailNotificationsEnabled: {
      type: Boolean,
      default: true,
    },
    isPushNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isMonthlyNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isQuarterNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
  },
  phone: {
    type: String,
    default: '',
  },
  socials: {
    twitter: {
      type: String,
      default: '',
    },
  },
  location: {
    type: String,
    default: '',
  },
  bio: {
    type: String,
    default: '',
  },
});

UserSchema.pre<IUserSchema>('save', function(next: HookNextFunction) {
  const user = this;
  if (this.isModified('password') || this.isNew) {
    bcrypt.genSalt(10, function(err, salt) {
      if (err) {
        return next(err);
      }

      bcrypt.hash(user.password, salt, function(err, hash) {
        if (err) {
          return next(err);
        }

        user.password = hash;
        next();
      });
    });
  } else {
    return next();
  }
});

UserSchema.methods.comparePassword = async function(password: IUserSchema['password']): Promise<Boolean> {
  const user = this;
  const match = await bcrypt.compare(password, user.password);
  return match;
};

UserSchema.methods.forgotPassword = async function(): Promise<SentMessageInfo> {
  const token = nanoid(72);

  return await new Promise((resolve, reject) => {
    sendResetMail(this.email, token)
      .then(async (info) => {
        await PasswordReset.create({
          email: this.email,
          token,
          createdAt: new Date(),
        });
        resolve(info);
        consola.success({
          badge: true,
          message: `Reset mail successfully sent: ${info}`,
        });
      })
      .catch((error) => {
        reject(error);
        consola.error({
          badge: true,
          message: `Error during sending reset email: ${error}`,
        });
      });
  });
};

export default model<IUserSchema>('User', UserSchema);
