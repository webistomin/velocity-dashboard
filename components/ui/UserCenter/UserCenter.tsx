import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { State } from 'vuex-class';

import BaseThumbnail from 'components/base/BaseThumbnail';
import BaseSidebar from 'components/base/BaseSidebar';
import BaseOverlay from 'components/base/BaseOverlay';
import UserProfile, { IUserProfileInfo } from 'components/ui/UserCenter/UserProfile/UserProfile';
import { IUserInterfaceDB } from 'common/types/user/user-schema';
import { UserRoles } from 'common/types/user/user-roles';

import './UserCenter.sass';

@Component({
  name: 'UserCenter',
})
export default class UserCenter extends VueComponent {
  public isUserProfileVisible: boolean = false;
  public userInfo: IUserProfileInfo = {
    firstName: '',
    lastName: '',
    email: '',
    location: '',
    bio: '',
    avatar: '',
    role: UserRoles.OPERATOR,
  };

  @State((state) => state.auth.user)
  getAuthUser!: IUserInterfaceDB;

  public toggleUserProfileVisibility() {
    this.isUserProfileVisible = !this.isUserProfileVisible;
  }

  public get shouldLockBody(): boolean {
    return this.isUserProfileVisible && window.matchMedia('(max-width: 1023px)').matches;
  }

  public onClickOutside(): void {
    this.isUserProfileVisible = false;
  }

  public toggleModalVisibilityByKeyDown(event: KeyboardEvent) {
    if (this.isUserProfileVisible && event.key === 'Escape') {
      this.onClickOutside();
    }
  }

  public mounted(): void {
    document.addEventListener('keydown', this.toggleModalVisibilityByKeyDown);
  }

  public beforeDestroy(): void {
    document.removeEventListener('keydown', this.toggleModalVisibilityByKeyDown);
  }

  public render(): VNode {
    const authUser = this.getAuthUser;
    this.userInfo.firstName = authUser.firstName;
    this.userInfo.lastName = authUser.lastName;
    this.userInfo.email = authUser.email;
    this.userInfo.location = authUser.location;
    this.userInfo.bio = authUser.bio;
    this.userInfo.avatar = authUser.avatar;
    this.userInfo.role = authUser.role;

    return (
      <div class='user-center' v-scroll-lock={this.shouldLockBody} v-click-outside={this.onClickOutside}>
        <button class='user-center__btn btn' onClick={this.toggleUserProfileVisibility}>
          <BaseThumbnail image={authUser.avatar} size='s' alt='User' />
        </button>

        <BaseSidebar isVisible={this.isUserProfileVisible} class='user-center__sidebar'>
          <UserProfile class={'user-center__profile'} info={this.userInfo} />
        </BaseSidebar>

        <BaseOverlay class='user-center__overlay' isVisible={this.isUserProfileVisible} onClick={this.onClickOutside} />
      </div>
    );
  }
}
