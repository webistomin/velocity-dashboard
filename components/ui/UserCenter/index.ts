import UserCenter from './UserCenter';

export default UserCenter;
