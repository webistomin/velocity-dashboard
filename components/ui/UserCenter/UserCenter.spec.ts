import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
// @ts-ignore
import vClickOutside from 'v-click-outside';
import VScrollLock from 'v-scroll-lock';
import { UserRoles } from 'common/types/user/user-roles';
import UserCenter from './UserCenter';

describe('UserCenter UI', () => {
  const localVue = createLocalVue();
  localVue.use(vClickOutside);
  localVue.use(VScrollLock);
  localVue.use(Vuex);
  let store: any;
  let state;

  beforeEach(() => {
    state = {
      user: {
        firstName: 'Alexey',
        lastName: 'Istomin',
        email: 'mail@mail.ru',
        location: '',
        bio: '',
        avatar: '',
        role: UserRoles.ADMIN,
      },
    };

    store = new Vuex.Store({
      modules: {
        auth: {
          namespaced: true,
          state,
        },
      },
    });
  });

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: jest.fn(() => {
        return {
          matches: true,
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
        };
      }),
      writable: true,
    });
  });

  const factory = (propsData = {}, options = {}) => {
    return shallowMount(UserCenter, {
      localVue,
      store,
      propsData: {
        ...propsData,
      },
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Toggle user center visibility', () => {
    const wrapper = factory();
    wrapper.setData({
      isUserProfileVisible: false,
    });
    wrapper.vm.toggleUserProfileVisibility();
    expect(wrapper.vm.isUserProfileVisible).toEqual(true);
  });

  it('Close modal on "Esc"', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).toHaveBeenCalled();
  });

  it('Delete global event listeners on destroy', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.destroy();
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).not.toHaveBeenCalled();
  });
});
