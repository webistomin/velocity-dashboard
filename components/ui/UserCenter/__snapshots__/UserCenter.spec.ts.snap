// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCenter UI Matches snapshot 1`] = `
"<div class=\\"user-center\\"><button class=\\"user-center__btn btn\\">
    <vuecomponent-stub image=\\"\\" alt=\\"User\\" size=\\"s\\"></vuecomponent-stub>
  </button>
  <vuecomponent-stub class=\\"user-center__sidebar\\">
    <userprofile-stub info=\\"[object Object]\\" class=\\"user-center__profile\\"></userprofile-stub>
  </vuecomponent-stub>
  <baseoverlay-stub class=\\"user-center__overlay\\"></baseoverlay-stub>
</div>"
`;
